lib/auth/ec2_join_test.go

```diff
@@ -0,0 +1,402 @@
+				},
+			},
+			ec2Client: ec2ClientRunning{},
+			request: RegisterUsingTokenRequest{
+				Token:               "test_token",
+				NodeName:            "node_name",
+				Role:                types.RoleNode,
+				HostID:              instance2.account + "-" + instance2.instanceID,
+				EC2IdentityDocument: instance2.iid,
+			},
+			expected: check.NotNil,
+			clock:    clockwork.NewFakeClockAt(instance2.pendingTime),
+		},
+		{
+			desc: "instance already joined, fake ID",
+			tokenRules: []*types.TokenRule{
+				&types.TokenRule{
+					AWSAccount: instance2.account,
+					AWSRegions: []string{instance2.region},
+				},
+			},
+			ec2Client: ec2ClientRunning{},
+			request: RegisterUsingTokenRequest{
+				Token:               "test_token",
+				NodeName:            "node_name",
+				Role:                types.RoleNode,
+				HostID:              "fake id",
+				EC2IdentityDocument: instance2.iid,
+			},
+			expected: check.NotNil,
+			clock:    clockwork.NewFakeClockAt(instance2.pendingTime),
+		},
+		{
+			desc: "instance not running",
+			tokenRules: []*types.TokenRule{
+				&types.TokenRule{
+					AWSAccount: instance1.account,
+					AWSRegions: []string{instance1.region},
+				},
+			},
+			ec2Client: ec2ClientNotRunning{},
+			request: RegisterUsingTokenRequest{
+				Token:               "test_token",
+				NodeName:            "node_name",
+				Role:                types.RoleNode,
+				HostID:              instance1.account + "-" + instance1.instanceID,
+				EC2IdentityDocument: instance1.iid,
+			},
+			expected: check.NotNil,
+			clock:    clockwork.NewFakeClockAt(instance1.pendingTime),
+		},
+		{
+			desc: "instance not exists",
+			tokenRules: []*types.TokenRule{
+				&types.TokenRule{
+					AWSAccount: instance1.account,
+					AWSRegions: []string{instance1.region},
+				},
+			},
+			ec2Client: ec2ClientNoInstance{},
+			request: RegisterUsingTokenRequest{
+				Token:               "test_token",
+				NodeName:            "node_name",
+				Role:                types.RoleNode,
+				HostID:              instance1.account + "-" + instance1.instanceID,
+				EC2IdentityDocument: instance1.iid,
+			},
+			expected: check.NotNil,
+			clock:    clockwork.NewFakeClockAt(instance1.pendingTime),
+		},
+		{
+			desc: "TTL expired",
+			tokenRules: []*types.TokenRule{
+				&types.TokenRule{
+					AWSAccount: instance1.account,
+					AWSRegions: []string{instance1.region},
+				},
+			},
+			ec2Client: ec2ClientRunning{},
+			request: RegisterUsingTokenRequest{
+				Token:               "test_token",
+				NodeName:            "node_name",
+				Role:                types.RoleNode,
+				HostID:              instance1.account + "-" + instance1.instanceID,
+				EC2IdentityDocument: instance1.iid,
+			},
+			expected: check.NotNil,
+			clock:    clockwork.NewFakeClockAt(instance1.pendingTime.Add(5*time.Minute + time.Second)),
+		},
+	}
+	for _, tc := range testCases {
+		println("Running test case:", tc.desc)
```

Comment:
Use subtests if you can; they give much better visibility in the test report about what's failing and why.

**EDIT:** Oh, you're stuck with a `Check` suite. Oh, well...

Reply:
Yeah I was lazy and used the Check suite, but realized it is pretty easy to just create my own auth server so went back and did that, and updated this test